addons/godot-xr-tools/functions/Function_JumpDetect_movement.gd

```diff
@@ -0,0 +1,154 @@
+##     Y velocity (in the tracking space) into a sliding-window averager. If both
+##     average Y velocities exceed a threshold parameter then the player has
+##     jumped. 
+##
+
+
+## Movement provider order
+export var order := 20
+
+## Enable detecting of jump via body (through the camera)
+export var body_jump_enable := true
+
+## Body jump detection threshold (M/S^2)
+export var body_jump_threshold := 2.5
+
+## Enable detectionm of jump via arms (through the controllers)
+export var arms_jump_enable := false
+
+## Arms jump detection threshold (M/S^2)
+export var arms_jump_threshold := 10.0
+
+
+# Sliding Average class
+class SlidingAverage:
+	# Sliding window size
+	var _size: int
+
+	# Sum of items in the window
+	var _sum := 0.0
+
+	# Position
+	var _pos := 0
+
+	# Data window
+	var _data := Array()
+
+	# Constructor
+	func _init(var size: int):
+		# Set the size and fill the array
+		_size = size
+		for i in size:
+			_data.push_back(0.0)
+
+	# Update the average
+	func update(var entry: float) -> float:
+		# Add the new entry and subtract the old
+		_sum += entry
+		_sum -= _data[_pos]
+
+		# Store the new entry in the array and circularly advance the index
+		_data[_pos] = entry;
+		_pos = (_pos + 1) % _size
+
+		# Return the average
+		return _sum / _size
+
+
+# Node Positions
+var _camera_position := 0.0
+var _controller_left_position := 0.0
+var _controller_right_position := 0.0
+
+# Node Velocities
+var _camera_velocity := SlidingAverage.new(5)
+var _controller_left_velocity := SlidingAverage.new(5)
+var _controller_right_velocity := SlidingAverage.new(5)
+
+# Node references
+onready var _origin_node := ARVRHelpers.get_arvr_origin(self)
+onready var _camera_node := ARVRHelpers.get_arvr_camera(self)
+onready var _controller_left_node := ARVRHelpers.get_left_controller(self)
+onready var _controller_right_node := ARVRHelpers.get_right_controller(self)
+
+
+# Perform jump detection
+func physics_movement(delta: float, player_body: PlayerBody):
+	# Handle detecting body jump
+	if body_jump_enable:
+		_detect_body_jump(delta, player_body)
+	
+	# Handle detecting arms jump
+	if arms_jump_enable:
+		_detect_arms_jump(delta, player_body)
+
+
+# Detect the player jumping with their body (using the headset camera)
+func _detect_body_jump(delta: float, player_body: PlayerBody) -> void:
+	# Get the camera instantaneous velocity
+	var new_camera_pos := _camera_node.transform.origin.y
+	var camera_vel := (new_camera_pos - _camera_position) / delta
+	_camera_position = new_camera_pos
+
+	# Ignore zero moves (either not tracking, or no update since last physics)
+	if abs(camera_vel) < 0.001:
+		return;
+
+	# Clamp the camera instantaneous velocity to +/- 2x the jump threshold
+	camera_vel = clamp(camera_vel, -2.0 * body_jump_threshold, 2.0 * body_jump_threshold)
+
+	# Get the averaged velocity
+	camera_vel = _camera_velocity.update(camera_vel)
+
+	# Detect a jump
+	if camera_vel >= body_jump_threshold:
+		player_body.request_jump()
```

Comment:
If the player physically jumps, should we then also add velocity to the jump or in this case just detect the jump and signal?

Reply:
I think it depends on the type of game - if someone's developing a basketball game then very few players are going to be able to pull off the Michael Jordan 4 foot vertical jump without a little "VR assistance". I was testing by jumping on one of my demo scenes jump pads to launch myself ~8 meters into the air.

As such, I think the best we can do is give the option to the developer. I'll add an exported property flag for 'physical jump only' which will report the jump rather than requesting it.

Reply:
I'm still thinking of suppressing the report if the player isn't on the ground, or the ground is too steep to jump according to the ground-physics settings. It doesn't seem to make sense to report the signal and potentially fire jump sound-effects if the player is in the middle of gliding when they physically jump.

Reply:
The PlayerBody.request_jump now supports an optional parameter (configurable option in JumpDetect) which suppresses the GroundPhysics velocity boost, so the jump only comes from the physical players jump.

Reply:
Owh cool, I can see a jungle gym sort of demo where you can jump around on bouncepads and grab onto the ceiling or something :)